File: samples/downstream/c++/downstream-connector.cpp
#include <endian.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <string.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <time.h>
#include <unistd.h>
#include <string>
#include <vector>

const char *FIFO_LGR2MNG = "/var/run/intdash2/logger_100.tx";
const char *FIFO_MNG2LGR = "/var/run/intdash2/logger_100.rx";

static int g_sig_num = 0;

static void
sig_handler(int sig)
{
    g_sig_num = sig;
}

struct StringMessage
{
    unsigned int sec;
    unsigned int nsec;
    std::string id;
    std::string data;
};

static bool
parse_msg(std::vector<unsigned char> &pool, StringMessage &msg, std::string &err)
{
    // Parse message
    // See: https://docs.intdash.jp/manual/intdash-agent-developer-guide/latest/ja/intdash-agent-developer-guide-ja.pdf
    //
    // +---------------+---------------+---------------+---------------+
    // |    Type(01)   |                    Legnth                     |
    // +---------------+---------------+---------------+---------------+
    // |                           Time Sec                            |
    // +---------------+---------------+---------------+---------------+
    // |                           Time NSec                           |
    // +---------------+---------------+---------------+---------------+
    // |     DType     |      Seq      |   ID Length   |      ID
    // +---------------+---------------+---------------+-----------//--+
    // |   Data String
    // +-----------//--+

    // Initialize

    err.erase();

    // 1(Type) + 3(Length) + 4(Time Sec) + 4(Time NSec) + 1(DType) + 1(Seq)
    const size_t minimum_msg_size = 1 + 3 + 4 + 4 + 1 + 1;
    // minimum_msg_size + 1(ID Length)
    const size_t offset_to_id = minimum_msg_size + 1;

    // Check size of pool

    if (pool.size() < minimum_msg_size)
    {
        return false;
    }

    //
    // Type must be 1 (0x01)
    //

    if (pool[0] != 1)
    {
        pool.erase(pool.begin(), pool.begin() + 1);
        err = "unexpected data: type must be 1";
        return false;
    }

    //
    // Length from Time Sec to end of Data String
    //

    size_t length_from_time_to_end = 0;
    length_from_time_to_end |= pool[1];
    length_from_time_to_end |= pool[2] << 8;
    length_from_time_to_end |= pool[3] << 16;

    size_t msg_len = 1 + 3 + length_from_time_to_end;

    // Check data size
    if (pool.size() < msg_len)
    {
        return false;
    }

    //
    // Time Sec & Time Nsec
    //

    msg.sec = le32toh(*(unsigned int *)(&pool[4]));
    msg.nsec = le32toh(*(unsigned int *)(&pool[8]));

    //
    // DType of String must be 29 (0x1D)
    // NOTE:
    //   Please note that intdash Edge Agent's DType of String (29, 0x1D)
    //   and intdash's Data Type of String (10, 0x0A) are different.
    //

    if (pool[12] != 0x1d)
    {
        pool.erase(pool.begin(), pool.begin() + msg_len);
        err = "dtype is not string";
        return false;
    }

    //
    // ID Length
    //

    size_t id_len = pool[14];

    //
    // ID
    //

    msg.id.assign(pool.begin() + offset_to_id, pool.begin() + offset_to_id + id_len);

    //
    // Data String
    //

    size_t offset_to_data = offset_to_id + id_len;
    size_t data_len = msg_len - offset_to_data;

    msg.data.assign(pool.begin() + offset_to_data, pool.begin() + offset_to_data + data_len);

    pool.erase(pool.begin(), pool.begin() + msg_len);
    return true;
}

int main(int argc, char *argv[], char *envp[])
{
    // Setup signal handler

    struct sigaction sa;
    sa.sa_handler = sig_handler;
    sigemptyset(&(sa.sa_mask));
    sa.sa_flags = 0;

    signal(SIGPIPE, SIG_IGN);
    sigaction(SIGINT, &sa, NULL);
    sigaction(SIGTERM, &sa, NULL);

    // Open FIFO

    int fd = open(FIFO_MNG2LGR, O_RDONLY);
    if (fd == -1)
    {
        fprintf(stderr, "fopen(%s) failed : %s (%d)\n", FIFO_MNG2LGR, strerror(errno), errno);
        return 1;
    }

    // Read data

    std::vector<unsigned char> pool;
    while (g_sig_num == 0)
    {
        // Parse data

        StringMessage msg;
        std::string err;
        if (parse_msg(pool, msg, err))
        {
            // Print message
            fprintf(stdout, "Timestamp: %u.%09u, id: %s, data: %s\n", msg.sec, msg.nsec, msg.id.c_str(), msg.data.c_str());
            continue;
        }
        if (err.length() > 0)
        {
            fprintf(stderr, "parse failed: %s\n", err.c_str());
            continue;
        }

        // Read data from FIFO

        unsigned char buf[minimum_msg_size];
        ssize_t readsize = read(fd, buf, sizeof(buf));
        if (readsize == -1)
        {
            fprintf(stderr, "read() failed: %s (%d)\n", strerror(errno), errno);
            break;
        }

        // Store data to pool

        pool.insert(pool.end(), buf, buf + readsize);
    }

    // Close FIFO

    close(fd);

    return 0;
}

File: samples/downstream/python/downstream-connector.py
#!/usr/bin/env python3
# coding: utf-8

import os
import select
import struct
import signal

FIFO_LGR2MNG = "/var/run/intdash2/logger_100.tx"
FIFO_MNG2LGR = "/var/run/intdash2/logger_100.rx"

SIGNUM = 0

def signalHandler(signum, frame):
  global SIGNUM
  SIGNUM = signum

# Packat format (type: String)
#
# +---------------+---------------+---------------+---------------+
# |    Type(01)   |                    Legnth                     |
# +---------------+---------------+---------------+---------------+
# |                           Time Sec                            |
# +---------------+---------------+---------------+---------------+
# |                           Time NSec                           |
# +---------------+---------------+---------------+---------------+
# |     DType     |      Seq      |   ID Length   |      ID
# +---------------+---------------+---------------+-----------//--+
# |   Data String
# +-----------//--+

class StringPacket():
  def __init__(self, pkt) -> None:
    offset = 0

    # Type
    type_size = 1
    (self.__type,) = struct.unpack_from('B', pkt, offset)
    offset += type_size

    # Length
    pkt_len_size = 3
    (pkt_len, ) = struct.unpack_from('<3s', pkt, offset)
    pkt_len += bytes([0])
    (self.__pkt_len,) = struct.unpack('<L', pkt_len)
    offset += pkt_len_size

    # Time Sec & Nsec
    sec_size = 4
    (self.__sec,) = struct.unpack_from('<L', pkt, offset)
    offset += sec_size
    nsec_size = 4
    (self.__nsec,) = struct.unpack_from('<L', pkt, offset)
    offset += nsec_size

    # DType
    dtype_size = 1
    (self.__dtype,) = struct.unpack_from('B', pkt, offset)
    offset += dtype_size

    # Seq
    seq_size = 1
    (self.__seq,) = struct.unpack_from('B', pkt, offset)
    offset += seq_size

    # ID Length & ID
    id_len_size = 1
    (self.__id_len,) = struct.unpack_from('B', pkt, offset)
    offset += id_len_size
    (self.__id,) = struct.unpack_from(str(self.__id_len) + 's', pkt, offset)
    offset += self.__id_len

    # Calc Header Length
    header_size = sec_size + nsec_size + dtype_size + seq_size

    # Data String
    data_len = self.__pkt_len - header_size - id_len_size - self.__id_len
    (self.__data,) = struct.unpack_from(str(data_len) + 's', pkt, offset)
    offset += data_len

    self.__pkt_size = offset

  @property
  def size(self):
    return self.__pkt_size

  @property
  def type(self):
    return self.__type

  @property
  def sec(self):
    return self.__sec

  @property
  def nsec(self):
    return self.__nsec

  @property
  def seq(self):
    return self.__seq

  @property
  def dtype(self):
    return self.__dtype

  @property
  def id(self):
    return self.__id.decode()

  @property
  def data(self):
    return self.__data


if __name__ == '__main__':
  # Setup signal handler
  signal.signal(signal.SIGINT, signalHandler)
  signal.signal(signal.SIGTERM, signalHandler)

  while SIGNUM == 0:
    # Open FIFO
    with os.fdopen(os.open(FIFO_MNG2LGR, os.O_RDONLY | os.O_NONBLOCK), 'rb') as fifo:
      # Main Loop
      while SIGNUM == 0:
        # Wait Data
        rready, _, _ = select.select([fifo], [], [], 1)
        if SIGNUM != 0:
          break
        if not rready:
          continue

        # Read Data
        data = fifo.read()
        while len(data) != 0:
          # Parse Data
          pkt = StringPacket(data)
          print('Timestamp: {0}.{1:0=9}, id: {2}, data: {3}'.format(pkt.sec, pkt.nsec, pkt.id, pkt.data))

          # Next Data
          tempdat = bytearray(data)
          del tempdat[:pkt.size]
          data = bytes(tempdat)

